### PR TITLE
Documentation cleanup for activation functions

### PR DESCRIPTION
**Pull request by pmitros**
This pull request improves documentation for the activation functions. 

Specifically, it:

* Splits activation functions into those which are a weighed sum+nonlinearity and the others
* Adds an image for each activation function so one can tell at-a-glance what it does
* Adds documentation for RReLU (which was previously undocumented)
* Adds default parameters for 'Threshold' so that optimizers may instantiate it in the same way as any other activation function (and so we have a default for what to plot)

Potential issues:

* This is a first pull request to this project. I am unfamiliar with community conventions
* I seem to be the first person adding images to docs. Are these where you'd like them?
* Are the sizes appropriate? Should I be concerned about interactions with specific IDEs?
* Are there better labels for the section headers? I considered using the language of generalized linear models and more background, but decided jargon may be less helpful to less experienced users

### CODE REVIEW

**Comment by pmitros** [comment]
This is an example of how the documentation renders in `build html`:

image

**Comment by apaszke** [comment]
That looks great, but I think it would be much better to add scripts that can be used to generate the images at build time instead of checking them in to the repo as they are now. Can you please try that? This has a few benefits, like allowing us to change the style of the plots in a very simple way. Thanks!

**Comment by soumith** [comment]
great improvements, thanks a lot.

I dont think adding scripts used to generate the images at build-time is THAT necessary. If you have them lying around, it'd be great to check them in, but if not we'll make do what the images as-is.

**Comment by vishwakftw** [comment]
Thanks for the PR. This helps with https://github.com/pytorch/pytorch/issues/5432 and the PR https://github.com/pytorch/pytorch/pull/5450.

**Comment by pmitros** [comment]
Thank you all for the feedback. I checked the script in.

Being new to the project, I'm not familiar with build conventions. In particular, please let me know whether:

* There is a better place for the script (I created a new directory `build/scripts`)
* Whether the images should build directly to `docs/build` and/or be part of the `Makefile` process.

I personally prefer both the simplicity of just having the script but still having the images check in and not adding another build step:

* For new developers, it's helpful to have this sort of documentation when browsing on `github`
* The script takes a moment to run, and I'd rather not slow down the build process
* The Makefile is very simple right now, and these are the sorts of things which end up complicating it as they build up
* The script depends on `pylab,` and I don't know whether we want the extra dependency

**Comment by apaszke** [comment]
Thanks a lot, this will be very useful.
* `source/scripts` should work fine. 
* no, let's keep them generated just like they are now (so they should run before the docs are built). It should be easy to extend the `Makefile` to run the python script before it runs `sphinx` (I think it's enough to add it [before this line](https://github.com/pytorch/pytorch/blob/master/docs/Makefile#L27)).

It's generally best to avoid checking in images, because git is pretty bad at handling binary data (the repo size gets much larger, and everyone has to download the images).

* It's not this bad, this part is executed only when a commit is pushed to master, and shouldn't bottleneck the infra (as long as it finishes within let's say a minute)
* It's just a single line that has to be added to Makefile
* PyTorch can be still built without `pylab`, it's just for the docs. Few people are building the docs locally, since they're generally available online.

**Comment by pmitros** [comment]
I removed the images and added to the `Makefile`, as per your preferences. I modified the build scripts to skip over images which had already been built, which is good practice regardless, so it shouldn't add much to time.

Regarding building docs locally, I'm more concerned about any local tools which interpret docstrings (IDEs, etc.). It'd be nice to have that just work. But half a meg is a decent chunk to add to a git repo, and that can grow if we decide to add illustrations more places. I take your point, though. `git` is surprisingly bad at handling bigger files ("surprisingly" since the data structures it uses scale to big files just fine).

For reference:

* The images, uncompressed, are 525k
* The images, recompressed losslessly with trimage, are 352k. Lossy compression depends on the setting
* The repository is 40M for pytorch, 52M with submodules
* The repository once built is 2.6G
* Rebuilding the images takes 3.5 seconds
* Skipping over built images takes 1 second (load Python imports)
* Rebuilding HTML pages without images takes 11.8 seconds

**Comment by SsnL** [comment]
Oh I didn't see this PR and wrote the RReLU doc myself in #5459 . I can remove my changes so these  two PRs won't conflict (hopefully). However, I have two comments on the RReLU doc:

1. it should be ax rather than x/a.
2. nit: there should be a comma after equation and ``where'' should be lower case.

**Comment by SsnL** [comment]
Also this is great improvement, thanks!

**Comment by pmitros** [comment]
@SsnL Good catch! That could have been embarrassing. I took formula from a source which divided by three rather than multiplying by 1/3. Fixed. "Where" is lower-case.

The comma, in the interests of consensus, I added as you suggested. I think it is clearer without (and most parts of the documentation appear to omit it). I added on the second line, since otherwise it typeset very poorly (vertical center and far-right)

File: torch/nn/modules/activation.py [hunk 1]
```diff
@@ -32,7 +34,7 @@ class Threshold(Module):
         >>> print(m(input))
     """
 
-    def __init__(self, threshold, value, inplace=False):
+    def __init__(self, threshold=0.0, value=0.0, inplace=False):
```
Comment (apaszke) [inline torch/nn/modules/activation.py]: Hmm can we keep the interface as it was? The default one is just ReLU, and it's clearer to use that module instead of `nn.Threshold()`

**Review by apaszke — COMMENTED** [review]
Looks great! Can you please revert the change of signature for `nn.Threshold`? Will be good to merge after that

**Comment by SsnL** [comment]
@pmitros Looks great, thanks!

File: torch/nn/modules/activation.py (hunk 1 above)
Comment (pmitros) [reply torch/nn/modules/activation.py]: This one, I'll push back on a little bit harder. It's important to be consistent on function signatures, especially in a language with duck typing, and especially in a field where one might want to introspect code to e.g. optimize hyperparameters or visualize what's going on.

Before this change, the nonlinear activation functions had different function signatures. In particular, it was impossible for a program to call `Threshold` without human intervention since there was no reasonable set of default values provided.

Use cases:

* The script which generated the images is oblivious to anything about the functions other that they are a nonlinear functions with one input and one output
* Optimizing hyperparameters, for example picking choice of nonlinearity through something like simulated annealing or genetic algorithms (and further optimizing parameter choices on those nonlinearities from a reasonable starting-point if desired).
* For learning purposes, having tools where students might have a visualization of these functions, and tweak parameters.

The next step would be to register these in some way (e.g. Python entry points).

Comment (pmitros) [reply torch/nn/modules/activation.py]: That said, I agree a human user should use `ReLU` when that's what's meant. I guess I have faith in people to do that without enforcing it :)

**Comment by pmitros** [comment]
@SsnL I looked over your PR. On the whole, your text is better than mine. I apologize for not noticing you were working on it. I merged the text. Please feel free to tweak further in your PR.

**Comment by SsnL** [comment]
@pmitros It was my bad. :) Thanks for merging the text. I'll rebase my PR on top of yours when it is merged.

File: torch/nn/modules/activation.py (hunk 1 above)
Comment (SsnL) [reply torch/nn/modules/activation.py]: These aren't really reasonable default values anyways. And I don't think I get the points 2 and 3. For 1, I would much rather change the script (e.g. maybe make a dict from function name to args, and print args in plot title), than change the signature of this.

Comment (apaszke) [reply torch/nn/modules/activation.py]: I agree with @SsnL. Threshold is really a family of functions, and you need to choose the one you want explicitly. If you want the defaults (which are indeed commonly used), we picked it already for you under the name of `nn.ReLU`. In the hyperparameter optimization case, if you want to limit yourself to parameter-less nonlinearities, you should just exclude Threshold in favor of ReLU.

Comment (pmitros) [reply torch/nn/modules/activation.py]: Thank you.

I removed Threshold from the PR. Given that the goals of documentation, I don't think an image with unreasonable defaults makes sense to include. My preference would be to add reasonable defaults for Threshold, but I think that's a separate discussion and PR, and a more limited change might makes more sense in this one.

I will respond to feedback to specific comments independently, since it may be moot.

Comment (pmitros) [reply torch/nn/modules/activation.py]: I'm responding to points above, but all of this is likely academic at this point, so please feel free to skip.

re: Points 2, 3.

In retrospect, that's clearly outside of the scope of this PR. It was a mistake on my part to raise them here. I'm glad to go into more depth if desired.

re: `Threshold`.

I think a key issue is that the terminology might be a little ambiguous. In classical neural networks literature, 'Threshold' generally referred to a step function:

image

([source](http://staff.itee.uq.edu.au/janetw/cmc/chapters/Introduction/): a random text from 1997)

Other [references](http://slideplayer.com/slide/4799311/) refer to *any* non-linear activation as a threshold function. At the time, thresholds generally were, as you stated, non-zero. A Google Images search 'threshold function deep learning' and similar brings up zero plots as used in pytorch, so this is likely to be a point-of-confusion.

Ideally, the documentation might explain:

* What the function does at-a-glance;
* Points of confusion around terminology;
* Why it does what it does; and
* When it ought to be used

I think that's especially important with a function like this one, where:

* in networks with bias terms, `Threshold` might often behave just like a `ReLU` (perhaps with slightly different initial conditions)
* there is a historical body of texts using the terminology differently

The target of the documentation are people who might not already be familiar with this.

re: Plots with args

My goal with the plots was to give information-at-a-glance. With that, I prefer less clutter. If I were putting more time into this, I'd probably go in the other direction:

* Shrink the plots a little bit;
* Add x=0, y=0 axes (faint);
* Reduce the number of labelled ticks; and
* Possibly, remove title entirely, since it is clear from the context

Alternatively, I might clearly label the parameters on the image (e.g. an arrow pointing to the threshold, labelling it as such).

If more clarity on how parameters influence things is desired, interactives where one can play with the parameters might be the next step.

**Comment by apaszke** [comment]
Thanks a lot @pmitros!
